bacon/model/Tags/TagManager.swift

```diff
@@ -0,0 +1,253 @@
+
+    /// Creates and returns a TagManager object.
+    /// - Note: A persistent TagManager is implemented as a singleton, and thus will never be outdated.
+    /// - Parameter withPersistence: If `true`, changes to TagManager data (e.g. add/remove Tags) will
+    ///     automatically be saved to storage. The TagManager object returned will also be pre-loaded
+    ///     with pre-existing data. If `false`, the TagManager object returned will only manage Tags in memory.
+    static func create(withPersistence: Bool) -> TagManager {
+        log.info("Creating TagManager withPersistence=\(withPersistence).")
+
+        // Without persistence
+        if !withPersistence {
+            log.info("Returning non-persistent instance of TagManager.")
+            return TagManager(withPersistence: false)
+        }
+
+        // With persistence: if a persistent TagManager has previously been instantiated,
+        // we return it. Otherwise, we create a new instance and cache it.
+        if let tagManager = TagManager.cachedPersistentTagManager {
+            log.info("Returning cached instance of persistent TagManager.")
+            return tagManager
+        }
+
+        log.info("No cached instance of persistent TagManager found. Instantiating TagManager.")
+        let fsm = FileStorageManager()
+        let tagManager: TagManager
+
+        do {
+            log.info("Reading TagManager data from storage.")
+            tagManager = try fsm.readFromJson(TagManager.self, file: TagManager.saveFileName)
+            log.info("TagManager reconstructed from storage data")
+        } catch {
+            // This could happen if a persistent TagManager is instantiated for the first time ever
+            log.warning("\(TagManager.errorLoadData). Instantiating a new instance.")
+            tagManager = TagManager(withPersistence: true)
+            TagManager.cachedPersistentTagManager = tagManager
+            return tagManager
+        }
+
+        log.info("Caching TagManager for future returns.")
+        TagManager.cachedPersistentTagManager = tagManager
+        return tagManager
+    }
+
+    // Private init to disallow direct instantiation: we want to adopt a semi-singleton design pattern.
+    // When TagManager is instantiated in non-persistent mode (e.g. for unit testing),
+    // we can create as many instances as possible.
+    // However, in normal operation (persistent mode), we want to allow only 1 instance,
+    // to avoid multiple working stores of tag data.
+    private init(withPersistence: Bool) {
+        isPersistent = withPersistence
+    }
+
+    func addChildTag(_ child: String, to parent: String) throws {
+        let childTag = Tag(child, parent: parent)
+        let parentTag = Tag(parent)
+
+        // childTag should not already exist
+        guard !allTags.contains(childTag) else {
+            throw DuplicateTagError(message: "\(childTag) already exists")
+        }
+
```

Comment:
Shall assert the parent tag already exists?

Reply:
I see that in the later part there is a method to add the parent as well when the parent is not already there. From the user side, he should never be able to add a child tag to a non-existent parent tag tho? if it happens it's probably some error within the programme?

Reply:
Okay, I'll change the specifications then.